### DISCUSSION TITLE
Migrating from LF

### DISCUSSION THREAD

**Discussion by idr4n** [sxyazi/yazi#848] [1. Q&A] [answered]
Hi and thanks for an amazing file manager. The preview implementation is superb compared to any other file managers I have used.

I'm currently trying to migrate my workflow from LF to Yazi. In LF, I have a shell script that saves a file/directory path to a specific file, which I then use with fzf to access those paths easily, sort of a bookmark or tag implementation. I know there is a bookmarks plugin, but to my knowledge they are not persistent yet across sessions.

So my questions is whether there is a good way to assign a multiple line shell snippet/command to a keymap (other than creating a separate executable shell script file). This is the snippet I would like to assign to a keymap:

```sh
    path="$1"
    file_tags="$HOME/Sync/file_tags.txt"

    # check if file_tags exists
    if [ ! -f $file_tags ]; then
        touch $file_tags
    fi

    # replace $HOME with ~ from the start of path
    path="${path/#$HOME/~}"

    if rg -q "^$path\$" $file_tags; then
        echo "Path already present."
    else
        # append the path to the file name and save it in the "file_tags" file
        echo "$path" >> $file_tags
    fi
```

Thanks in advance for your help!

**Comment by lpnh** [comment]
Hi! Sorry for not addressing your question directly, but have you considered the built-in [zoxide](https://github.com/ajeetdsouza/zoxide) integration in Yazi? Zoxide learns your frequently used directories, letting you jump to them easily –  it might offer a similar workflow to your fzf-based file tagging approach.

**Reply by idr4n** [reply]
Thanks for the suggestion. Yes I have. I use Zoxide in my workflow in general and with other scripts for Tmux session management but in this case it gives me too many directories and have to type too much to find the one I need as the directories I want to store and not so frequently visited. By having just a short list instead I can get where I want more easly (sort of a 'harpoon' workflow).

**Comment by sxyazi** [comment, marked as answer]
I think you can do it like this: 

```toml
[[manager.prepend_keymap]]
on   = [ "i" ]
run  = '''
shell --block --confirm '
    path="$1"
    file_tags="$HOME/Sync/file_tags.txt"

    # check if file_tags exists
    if [ ! -f $file_tags ]; then
        touch $file_tags
    fi

    # replace $HOME with ~ from the start of path
    path="${path/#$HOME/~}"

    if rg -q "^$path\$" $file_tags; then
        echo "Path already present."
    else
        # append the path to the file name and save it in the "file_tags" file
        echo "$path" >> $file_tags
    fi
'
'''
```

But this isn't very ergonomic, I would recommend implementing it through a plugin, if you need to hide Yazi and activate fzf, you can achieve this using the [`ya.hide()`](https://github.com/sxyazi/yazi/pull/792) API. 

BTW, I just merged a new [data distribution system](https://github.com/sxyazi/yazi/pull/826) that will allow cross-process communication and state synchronization through plugins, I think this will be useful for plugins like bookmarks.

**Reply by idr4n** [reply]
Thanks a lot for your reply!

Yes, that pretty much works as expected, but you are right, it is more like a hack than a proper tag or bookmark implementation. Furthermore. when I use fzf to search for the stored tags and call `yazi` with the selection, then obviously another nested yazi subprocess is created, which is far from ideal. In LF I could send the selection to the same instance, and therefore the new data distribution system that you just merged is pretty good addition, thank you for that. 

This is what I have for the moment until I try implementing it through a plugin as you mentioned, or the current bookmarks plugin is updated to make the bookmarks persistent across sessions :)

```
{ on = [ "b", "a" ], run = '''
shell --block --confirm '
  path="$1"
  tags_file="$HOME/Sync/tags_file.txt"

  # check if tags_file exists
  if [ ! -f $tags_file ]; then
    touch $tags_file
  fi

  # replace $HOME with ~ from the start of path
  path="${path/#$HOME/~}"

  if rg -q "^$path\$" $tags_file; then
    echo "Path already present."
  else
    # append the path to the file name and save it in the "tags_file" file
    echo "$path" >> $tags_file
  fi
'
''', desc = "Add tag" },

{ on = [ "b", "s" ], run = '''
shell --block --confirm '
sel=$(cat ~/Sync/tags_file.txt | sort | fzf --layout=reverse)
yazi --cwd-file ~/.cache/yazi/last_dir "$sel"
'
''', desc = "Search tag" }

```

**Reply by sxyazi** [reply]
If you want to avoid starting a new instance, you can achieve that by using [`ya.manager_emit("cd", { "/path/to/file" })`](https://yazi-rs.github.io/docs/plugins/utils#ya.manager_emit), which should be similar to sending a `cd` command to the current process like `lf` does.

The new DDS is designed for cross-process state synchronization, if your workflow doesn't require it and you just want to `cd` within the current instance, you can use this simpler approach. However, since it's a Lua API, you still need a plugin to implement it :D

**Reply by idr4n** [reply]
Great, that's good to know. I will give it a try hopefully soon! Btw, I was thinking if it would be a good approach when creating such plugin, to add a mark to the files that are in the tags file. Not sure whether this is currently possible in the API to modify the text or look of a particular path or item, but to plan ahead, I was thinking what would be the best way to store the tagged paths so when parsing and checking each item agains those tagged path is efficient, for example, using a key value lua table, in a tree structured kind of way:

```lua
{
  ["dir1"] = {
     "file1",
     "file2",
     ["dir2"] = {
        "file1"
     }
}
```

Thanks in advance!
